src/app/map/leaflet-map/leaflet-map.component.ts

```diff
@@ -0,0 +1,239 @@
+
+import 'leaflet';
+import 'leaflet.markercluster';
+import * as L from 'leaflet';
+
+declare module 'leaflet' {
+  export interface FeatureGroup<P = any> {
+    projectId: number;
+  }
+  export interface Marker<P = any> {
+    projectId: number;
+  }
+}
+
+const markerIconYellow = L.icon({
+  iconUrl: 'assets/images/marker-icon-yellow.svg',
+  // Retina Icon is not needed here considering we're using an SVG. Enable if you want to change to a raster asset.
+  // iconRetinaUrl: 'assets/images/marker-icon-2x-yellow.svg',
+  iconSize: [36, 36],
+  iconAnchor: [18, 36],
+  tooltipAnchor: [16, -28]
+});
+
+@Component({
+  selector: 'leaflet-map',
+  templateUrl: './leaflet-map.component.html',
+  styleUrls: ['./leaflet-map.component.scss']
+})
+
+export class LeafletMapComponent implements OnInit, AfterViewInit, OnChanges, OnDestroy {
+
+  @Input() project: Project;
+  @Input() zoom = 6;
+  @Input() thumbnail = false;
+
+  public loading = false;
+  readonly defaultBounds = L.latLngBounds([48, -139], [60, -114]); // all of BC
+  readonly maxBounds = L.latLngBounds([40, -150], [70, -110]); // all of BC
+  public projects: Array<Project> = [];
+  public selectedProject: Project = null;
+  private map: L.Map = null;
+  private markerList: Array<L.Marker> = [];
+  private markerClusterGroup = L.markerClusterGroup({
+    showCoverageOnHover: false,
+    maxClusterRadius: 40, // 0 disables
+  });
+
+  constructor(
+    private projectService: ProjectService,
+    private router: ActivatedRoute,
+    private appRef: ApplicationRef,
+    private injector: Injector,
+    private resolver: ComponentFactoryResolver
+  ) { }
+
+  ngOnInit() {
+    // Load the projects. If this is a thumbnail with a project set, ignore
+    if (!this.thumbnail) {
+
+      let routerProjId = this.router.snapshot.paramMap.get('project');
+
+      this.projectService.getAll().subscribe(results => {
+        if (results) {
+          this.projects = results;
+
+          // add data to projects layer
+          this.projects.forEach((proj, index) => {
+            if (proj.latitude && proj.longitude) {
+              const title = `Project: ${proj.name}`;
+              const marker = L.marker(L.latLng(proj.latitude, proj.longitude), { title: title })
+              .setIcon(markerIconYellow)
+              .on('click', L.Util.bind(this.onMarkerClick, this, proj));
+
+              marker.projectId = index;
+              this.markerList.push(marker); // save to list
+              this.markerClusterGroup.addLayer(marker); // save to marker clusters group
+
+              // did we navigate to the map with a poject defined?
+              // if so, zoom to the icon and open the popup
+              if (routerProjId && routerProjId === proj._id) {
+                this.selectedProject = proj;
+                this.map.setView(new L.LatLng(proj.latitude, proj.longitude), 10);
+
+                setTimeout(() => {
```

Comment:
Is this to wait for the map to finish setting the view?

Reply:
Yes. `setTimeout`, the duct tape of javascript. It shouldn't be needed, but in general calls to alter the map or invalidate the map in leaflet while the view is changing can cause wierdness with the display, which a timeout should prevent. Could likely be shorter than a second.

Reply:
Thought so. `setTimeout` always triggers me to double check ha!